### ISSUE TITLE
Ontology Information table: add rows for gene template and taxonomy

### ISSUE THREAD

**Issue by nataled** [PROconsortium/PRoteinOntology#162]
Create mock-ups for this table with rows added for gene template and taxon. These rows should only be displayed for terms that contain such information.

**Comment by nataled** [comment]
@karenross to mock up how to show these and to which categories these should apply.

**Comment by karenross** [comment]
Here is a mock-up
[ontology_info_mockup_Sept2020.pptx](https://github.com/PROconsortium/PRoteinOntology/files/5284767/ontology_info_mockup_Sept2020.pptx)

**Comment by nataled** [comment]
I think we should put the new lines between Parent and Terms by PRO Category, somewhat mimicking the order of an OBO file. Definitely not between name and synonyms, which should never be separated.

The has_gene_template line can begin with either 'relationship' or 'intersection_of', so just look for has_gene_template in the same way you say to look for only_in_taxon.

Other than the above, the only suggestion I have is to strip the organism out of the gene name (basically, remove the last parenthetical expression). I only added the organism name because of the ontological insistence that all names be unique (for example, rat and mouse genes would all be identical otherwise).

**Comment by Julie-Cowart** [comment]
Agree with Darren on ordering. Both the gene template and the taxon would
be hyperlinked to their appropriate external pages based on idspace.

*Gene Template*: HGNC:2514
<https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2514> CTNNB1
(human)
*Taxon*: NCBITaxon:9606 <http://purl.obolibrary.org/obo/NCBITaxon_9606> Homo
sapiens

As for the name to display, for taxons we have both the name and common
name. E.g for NCBITaxon:9606 the name is Homo sapiens and
the PRO-common-name synonym is human. We could use either (or both). The
gene template just has a name. E.g. the stanza for HGNC:2514 says it has a
name of "CTNNB1 (human)". It might cause bugs to assume anything in
parentheses should be thrown out so I'd just leave the name as is or add a
PRO-comon-name synonym to use instead that has the original name without
species added, e.g CTNNB1.

**Comment by karenross** [comment]
Revised mock-up. I changed the order of the rows according to Darren's and Julie's suggestion, I added the information about the hyperlinks, and indicated that any mention of "has_gene_template" should be grabbed (not just those that say "relationship: has_gene_template"). For the moment, I left the organism name in parentheses after the gene template name due to Julie's concern. I also left "Homo sapiens" as the taxon name (rather than human) just because that is currently how it is written in the stanza so it is the easiest thing to do. We can discuss whether we want to use Homo sapiens, human, or both.
[ontology_info_mockup_Sept2020.pptx](https://github.com/PROconsortium/PRoteinOntology/files/5285449/ontology_info_mockup_Sept2020.pptx)

**Comment by nataled** [comment]
The name given on the taxonomy line should be the official taxonomy name, in my opinion. That's also why I suggest to remove the organism name from the gene template line: it isn't actually part of the name. I do have a foolproof way of doing it if you need it, but it's written in perl. Basically, it splits a string value into an array of individual characters, reverses it, then retains characters that are within the outer set of parentheses. Reversing the result gives the organism name, which I then remove from the original name.

**Comment by karenross** [comment]
One thing I hadn't considered...
If there are multiple gene templates do we want them each on a separate line? Separated by ";"? Something else?

**Comment by nataled** [comment]
I think it'll be too busy to see if we have all on a line (or wrapped, as the case may be). I'd opt for one per line. Better yet, put each in a separate cell in the table. Will that work? Here's are examples of terms with multiple has_gene_template lines:

PR:Q9ULZ0 6 genes
PR:Q9JMA7 2 genes
PR:Q5JQC4 12 genes

**Comment by karenross** [comment]
I removed the organism name from the gene template and I added a slide with an example with multiple gene templates. I did each template on a separate line rather than in individual table cells. This is because the synonyms row has multiple lines but not individual table cells and I thought the template row looked better with the same style.
[ontology_info_mockup_Sept2020.pptx](https://github.com/PROconsortium/PRoteinOntology/files/5285846/ontology_info_mockup_Sept2020.pptx)

**Comment by nataled** [comment]
This is better, though I imagine Hongzhan will not like all the whitespace that will be created. Technically the synonyms are not on separate lines, just synonym types. So when you look at the second slide, the Gene-based synonyms are all on one line. It looks okay there because we aren't showing both the name and the identifier.

**Comment by karenross** [comment]
I don't have a good suggestion for the white space issue except for listing multiple template on the same line. Slide three shows what it would look like with a semi-colon separate list. I think the semi-colon version looks OK, and again, the number of cases where there will be more than two gene templates is very small.
[ontology_info_mockup_Sept2020.pptx](https://github.com/PROconsortium/PRoteinOntology/files/5286884/ontology_info_mockup_Sept2020.pptx)

**Comment by nataled** [comment]
I agree, it looks fine. We'll present these at the next PIR-PRO meeting for discussion. Thanks!

**Comment by Julie-Cowart** [comment]
I assume with this addition the gene_template terms should be removed from the Related Cross References table. They are currently listed in that section, all on one line (comma seperated and hyperlinked) and lumped in with any other xref terms.

See the 6 listed for https://proconsortium.org/app/entry/PR%3AQ9ULZ0/

**Comment by Julie-Cowart** [comment]
A preliminary version is on the test site. I have not yet stripped out the ' (human)' part from the name of the gene_template. 

@nataled what is the logic for striping the trailing organism name? You mentioned you had some code in perl? I need to write in python but it's using regular expressions it's basically the same. I could just strip off anything inside of parentheses (and the prefix space) but are there cases with parentheses in the name already? Am I over thinking?

**Comment by nataled** [comment]
The intended changes look good so far, though there seems to be one unintended change: each entry now has 'gene_templates' written just above the Ontology Information table.

As for the organism names, you aren't over thinking, as there are indeed parentheses in organism names on occasion. The way I remove them is by taking advantage of the fact that the organism-level terms have names that always end with a parenthesis, and always with the organism name indicated there. I reverse the name string, then delete all characters between that parenthesis and the matching one. The trick is finding the matching one. I use a count mechanism, like so:

blahblah protein (human)
 becomes
)namuh( nietorp halbhalb

I then count the closed and open parentheses--add one for each closed, subtract for each open--until the number reaches zero, and delete everything up to that point. So:
)  -> parenthesis_count=1, delete
n  count > 0 so delete
a  same
m  same
u  same
h  same
(  parenthesis_count = 0, stop process

The reversed string will now begin with a open parenthesis and a space, which are deleted before reversing string again.

A more complicated example: non-structural protein 1 (Influenza A virus (strain A/Brevig Mission/1/1918 H1N1))

reversed:
```
))1N1H 8191/1/noissiM giverB/A niarts( suriv A azneulfnI( nietorp larutcurts-non
122222222222222222222222222222222222211111111111111111110
```
Delete all characters for which the parenthesis count is greater that zero, stop process, delete first two characters (an open parenthesis and a space), then reverse string again. This will work even if the protein or gene name has a parenthesis in it.

**Comment by Julie-Cowart** [comment]
Ok I see and I can code that up. But if the taxon name is the only place we have to worry about parentheses in the string then we can do this simpler with regexp. The pattern '(.*?) \(.*\)' the first .* is reluctant and the second inside the match for the paren characters is greedy. Therefore the pattern will find the outermost set of parens (and space) and the matching group should be the gene name. This only  works if no gene names have the ( or ) character in them. Is that an assumption we can make? If not then yes I think we have to count match parens from the end (as long as we assume no unmatched parens in taxon names which really should be true).

**Comment by nataled** [comment]
I'm getting an error message when I click the link. Let me know if it is working for you. If it is, I'll send a screen shot of the error.

**Comment by Julie-Cowart** [comment]
It should be working at the moment. I'm working on another issue right now
and may have broken the test site temporarily.

**Comment by nataled** [comment]
Ah okay. Looks good so far, though I'm rather confused as to the source of 'flu-PA' in the note at the top of the entry page.

**Comment by Julie-Cowart** [comment]
I haven't gotten to changing that part yet so it's the parents' short name.
I believe after I fix issue #161 then it will be "This page represents a
class of proteins encompassing all the protein products of the *PA*
gene in *Influenza
A virus (A/Puerto Rico/8/1934(H1N1))*."

**Comment by nataled** [comment]
This change is good to go for production.
